### ISSUE TITLE
.curvature randomly being unknown

### ISSUE THREAD

**Issue by timothy-nunn** [cvxpy/cvxpy#2083]
**Describe the bug**
Curvature of the attached example is not deterministic. It appears to be coming from the `quad_form` routine which will sometimes be convex, and sometimes be unknown. 

**To Reproduce**
Please run attached python file (as a file because the matrix used is huge and I cannot replicate this issue for random numbers, for this I can only apologise). The gist is we create what should be a convex atom over and over again and print its curvature each time.

**Expected behavior**
The eigenvalues of `B` are all positive, so I would expect the `quad_form_part.curvature` line to print as "Convex" everytime. Regardless, I would expect consensus on whether the problem has unknown or convex curvature.

**Output**
```
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'UNKNOWN'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'UNKNOWN'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'UNKNOWN'
quad_form_part.curvature = 'CONVEX'
quad_form_part.curvature = 'UNKNOWN'
quad_form_part.curvature = 'CONVEX'
...
```

**Version**
 - OS: macOS Ventura 13.0.1 (Apple M1 MacBook Air)
 - CVXPY Version: 1.2.1 and 1.3.1
 - Python Version: 3.8.13
 - Scipy Version: 1.9.1
 - Numpy Version: 1.23.2
[non_deterministic_curvature.py.txt](https://github.com/cvxpy/cvxpy/files/11099460/non_deterministic_curvature.py.txt)

**Comment by rileyjmurray** [comment]
Thanks for reporting, @timothy-nunn. The problem is that we use the scipy method ``eigvalsh`` to try and compute the (algebraically) smallest eigenvalue of a given matrix. That function uses a random starting vector in an iterative method. We can fix this specific issue by generating the random starting vector for ``eigvalsh`` ourselves, with a fixed random seed. We'd welcome a contribution from you if you would be up for implementing that.

**Comment by Paulnkk** [comment]
@Transurgeon I checked for this issue, and the function eigvalsh itself (from my current understanding) does not provide input parameters to specify a starting vector for the algorithm which generates the eigenvalues. I do not know how to make this kind of changes since I assume you need to manipulate scipy itself and deploy this manipulated version in cvxpy

**Comment by Transurgeon** [comment]
Hi @Paulnkk, you are right. The [documentation](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.eigvalsh.html) for eigvalsh has no mention of an input parameter for specifying a starting vector. 
Perhaps we could take a look at the [source code](https://github.com/scipy/scipy/blob/v1.10.1/scipy/linalg/_decomp.py#L270-L620) for linalg.eigh. But if you are right about needing to manipulate scipy, I also have no clue on how to make those changes.
@rileyjmurray, could explain what you meant when you mention that the "the function uses a random starting vector in an iterative method"?

**Comment by SteveDiamond** [comment]
This probably won't work, but it might be worth trying setting the numpy random seed before calling eigh and seeing if that produces more consistent outputs.

**Comment by rileyjmurray** [comment]
Sorry for the confusion! I got the function name wrong. The correct function is 

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html

The relevant argument is ``v0``.

**Comment by Transurgeon** [comment]
Hey @rileyjmurray , I tried to add a simple random seed to choose a vector as the starting argument v0.
However, when running the pytest unit testing suite, I run into fails for these two methods.
<img width="1298" alt="image" src="https://user-images.githubusercontent.com/89562186/229004061-f675e601-7808-4c22-86ad-2fa0993f4ba0.png">
I did some digging and noticed that it only runs into issues when dealing with a sparse diagonal matrix. 
For example, in [test_constant_psd_nsd](https://github.com/cvxpy/cvxpy/blob/6d7e27bc0afddb51ae7c8ef21765a4d2932bf751/cvxpy/tests/test_expressions.py#L223), all asserts pass successfully until we start dealing with [sparse matrices](https://github.com/cvxpy/cvxpy/blob/6d7e27bc0afddb51ae7c8ef21765a4d2932bf751/cvxpy/tests/test_expressions.py#L272).
I saw some past similar issues, but im not sure if they are entirely related... 
Do you have any ideas why this is happening?

**Comment by Paulnkk** [comment]
@Transurgeon I created a PR for this issue and we can work on the branch together. I found another issue from my point of view, which I addressed in the PR, maybe you have an idea how to deal with it (it is not related to the issue you mentioned here, but more of an problem related to the convexity property itself for this case).

**Comment by rileyjmurray** [comment]
Resolved by #2086.
